From 7fa7d9966b81278fa8d7360b147ef3178be89b8c Mon Sep 17 00:00:00 2001
From: ChenXiaoSong <chenxiaosong@kylinos.cn>
Date: Wed, 17 Jun 2026 07:05:02 +0000
Subject: [PATCH 3/3] smb/server: support rename events notify

Example:

  smbclient //192.168.53.210/test -U root%1
  smb: \> notify /
  # server: touch /export/file1
  # server: mv /export/file1 /export/file2
  0004 file1
  0005 file2
  0003 file2
  0002 file1
  # server: mkdir /export/dir
  # server: mv file2 /export/dir/
  0002 file2
  # server: mv /export/dir/file2 /export/
  0001 file2

Signed-off-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/server/notify.c | 126 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 126 insertions(+)

diff --git a/fs/smb/server/notify.c b/fs/smb/server/notify.c
index c504a8f06ff0..36deefd8163e 100644
--- a/fs/smb/server/notify.c
+++ b/fs/smb/server/notify.c
@@ -24,6 +24,13 @@ struct ksmbd_notify_event {
 	char name[];
 };
 
+struct ksmbd_notify_moved_from {
+	u32 mask;
+	u32 cookie;
+	u32 namelen;
+	char name[];
+};
+
 struct ksmbd_notify_mark {
 	struct fsnotify_mark fsn_mark;
 };
@@ -35,6 +42,8 @@ struct ksmbd_notify_ctx {
 	struct ksmbd_work *work;
 	struct fsnotify_group *group;
 	struct ksmbd_notify_mark *inode_mark;
+	struct ksmbd_notify_moved_from *moved_from;
+	struct delayed_work moved_from_work;
 	u32 completion_filter;
 	u32 output_buffer_length;
 	bool events_ready;
@@ -115,6 +124,95 @@ static bool ksmbd_notify_filter_match(u32 filter, u32 mask, u32 mark_mask)
 	return ((filter & FILE_NOTIFY_CHANGE_FILE_NAME) != 0);
 }
 
+static void ksmbd_notify_trigger_orphaned_moved_from(struct ksmbd_notify_ctx *ctx)
+{
+	struct ksmbd_notify_moved_from *from = ctx->moved_from;
+	struct qstr name;
+
+	if (!from)
+		return;
+
+	name = QSTR_LEN(from->name, from->namelen);
+	if (ksmbd_notify_filter_match(ctx->completion_filter, from->mask,
+				      ctx->inode_mark->fsn_mark.mask))
+		ksmbd_notify_queue_event_under_lock(ctx, FILE_ACTION_REMOVED,
+						    &name);
+
+	kfree(from);
+	ctx->moved_from = NULL;
+}
+
+static void ksmbd_notify_moved_from_timeout(struct work_struct *work)
+{
+	struct ksmbd_notify_ctx *ctx = container_of(work, struct ksmbd_notify_ctx,
+						    moved_from_work.work);
+	unsigned long flags;
+
+	spin_lock_irqsave(&ctx->lock, flags);
+	ksmbd_notify_trigger_orphaned_moved_from(ctx);
+	spin_unlock_irqrestore(&ctx->lock, flags);
+
+	wake_up(&ctx->waitq);
+}
+
+static void ksmbd_notify_save_moved_from(struct ksmbd_notify_ctx *ctx, u32 mask,
+					 u32 cookie, const struct qstr *file_name)
+{
+	struct ksmbd_notify_moved_from *from;
+
+	cancel_delayed_work(&ctx->moved_from_work);
+
+	if (ctx->moved_from)
+		ksmbd_notify_trigger_orphaned_moved_from(ctx);
+
+	from = kmalloc(struct_size(from, name, file_name->len + 1), GFP_KERNEL);
+	if (!from)
+		return;
+
+	from->mask = mask;
+	from->cookie = cookie;
+	from->namelen = file_name->len;
+	memcpy(from->name, file_name->name, file_name->len);
+	from->name[file_name->len] = '\0';
+	ctx->moved_from = from;
+
+	schedule_delayed_work(&ctx->moved_from_work, msecs_to_jiffies(100));
+}
+
+static int ksmbd_notify_handle_local_rename(struct ksmbd_notify_ctx *ctx,
+					    u32 mask,
+					    const struct qstr *file_name)
+{
+	struct ksmbd_notify_moved_from *from = ctx->moved_from;
+	struct qstr from_name;
+
+	if (!from)
+		return 0;
+
+	from_name = QSTR_LEN(from->name, from->namelen);
+
+	if (ksmbd_notify_filter_match(ctx->completion_filter, from->mask,
+				      ctx->inode_mark->fsn_mark.mask))
+		ksmbd_notify_queue_event_under_lock(ctx,
+						    FILE_ACTION_RENAMED_OLD_NAME,
+						    &from_name);
+
+	if (ksmbd_notify_filter_match(ctx->completion_filter, mask,
+				      ctx->inode_mark->fsn_mark.mask))
+		ksmbd_notify_queue_event_under_lock(ctx,
+						    FILE_ACTION_RENAMED_NEW_NAME,
+						    file_name);
+
+	if (!(mask & FS_ISDIR) &&
+	    (ctx->completion_filter & FILE_NOTIFY_CHANGE_CREATION) &&
+	    ksmbd_notify_filter_match(ctx->completion_filter, FS_ATTRIB,
+				      ctx->inode_mark->fsn_mark.mask))
+		ksmbd_notify_queue_event_under_lock(ctx, FILE_ACTION_MODIFIED,
+						    file_name);
+
+	return 0;
+}
+
 static int ksmbd_notify_handle_event(struct fsnotify_group *group, u32 mask,
 				     const void *data, int data_type,
 				     struct inode *dir,
@@ -141,7 +239,19 @@ static int ksmbd_notify_handle_event(struct fsnotify_group *group, u32 mask,
 
 	spin_lock_irqsave(&ctx->lock, flags);
 
+	if (event_mask & FS_MOVED_FROM) {
+		if (file_name && file_name->len)
+			ksmbd_notify_save_moved_from(ctx, mask, cookie,
+						     file_name);
+		goto out_unlock;
+	}
+
 	if (event_mask & FS_MOVED_TO) {
+		if (ctx->moved_from && ctx->moved_from->cookie == cookie) {
+			ret = ksmbd_notify_handle_local_rename(ctx, mask,
+							       file_name);
+			goto out_unlock;
+		}
 		action = FILE_ACTION_ADDED;
 	} else if (event_mask & FS_CREATE) {
 		action = FILE_ACTION_ADDED;
@@ -156,6 +266,7 @@ static int ksmbd_notify_handle_event(struct fsnotify_group *group, u32 mask,
 		ret = ksmbd_notify_queue_event_under_lock(ctx, action,
 							  file_name);
 
+out_unlock:
 	spin_unlock_irqrestore(&ctx->lock, flags);
 	return ret;
 }
@@ -226,9 +337,16 @@ static void ksmbd_notify_unwatch(struct ksmbd_notify_ctx *ctx)
 {
 	struct fsnotify_group *group = ctx->group;
 	struct ksmbd_notify_mark *imark = ctx->inode_mark;
+	unsigned long flags;
 
 	ctx->watching = false;
 
+	cancel_delayed_work_sync(&ctx->moved_from_work);
+
+	spin_lock_irqsave(&ctx->lock, flags);
+	ksmbd_notify_trigger_orphaned_moved_from(ctx);
+	spin_unlock_irqrestore(&ctx->lock, flags);
+
 	if (!group)
 		return;
 
@@ -340,6 +458,7 @@ int smb2_setup_notify(struct ksmbd_work *work,
 	init_waitqueue_head(&ctx->waitq);
 	spin_lock_init(&ctx->lock);
 	INIT_LIST_HEAD(&ctx->events);
+	INIT_DELAYED_WORK(&ctx->moved_from_work, ksmbd_notify_moved_from_timeout);
 	ctx->work = work;
 	ctx->completion_filter = le32_to_cpu(req->CompletionFilter);
 	ctx->output_buffer_length = le32_to_cpu(req->OutputBufferLength);
@@ -377,9 +496,16 @@ int smb2_setup_notify(struct ksmbd_work *work,
 		work->send_no_response = 1;
 		rc = -ECANCELED;
 	} else if (ctx->events_ready) {
+		unsigned long flags;
+
 		/* Sleep waiting for other events to arrive. */
 		msleep(1000);
 
+		cancel_delayed_work_sync(&ctx->moved_from_work);
+		spin_lock_irqsave(&ctx->lock, flags);
+		ksmbd_notify_trigger_orphaned_moved_from(ctx);
+		spin_unlock_irqrestore(&ctx->lock, flags);
+
 		rc = smb2_build_notify_response(work, rsp, ctx);
 		if (rc) {
 			if (rc == -ENOSPC)
-- 
2.54.0

